istioctl/pkg/genbinding/genbinding.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package genbinding
+
+import (
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+type hostPort struct {
+	host string
+	port int
+}
+
+const defaultMultiMeshPort = "15443"
+
+var validHostnameRegex = regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
+
+// CreateBinding converts desired multicluster state into Istio state
+func CreateBinding(service string, clusters []string, labels map[string]string, egressGateway string, namespace string) ([]model.Config, error) { // nolint: lll
+
+	remoteService, err := parseNet(service, "")
+	if err != nil {
+		return nil, err
+	}
+
+	var remoteClusters []hostPort
+	for _, cluster := range clusters {
+		remoteCluster, err := parseNet(cluster, defaultMultiMeshPort)
+		if err != nil {
+			return nil, err
+		}
+		remoteClusters = append(remoteClusters, *remoteCluster)
+	}
+
+	istioConfig, err := serviceToServiceEntrySniCluster(*remoteService, remoteClusters, labels, egressGateway, namespace)
+	return istioConfig, err
+}
+
+// serviceToServiceEntry() creates a ServiceEntry pointing to istio-egressgateway
+func serviceToServiceEntrySniCluster(remoteService hostPort, remoteClusters []hostPort, labels map[string]string, egressGateway string, namespace string) ([]model.Config, error) { // nolint: lll
+	protocol := "http"
+	var egresslabels map[string]string
+	// It is strongly recommended addresses for different hosts don't clash;
+	// we use the hash to make clashing unlikely.
+	h := fnv.New32a()
+	h.Write([]byte(remoteService.host))
+	address := fmt.Sprintf("127.255.%d.%d", h.Sum32()&255, (h.Sum32()>>8)&255)
+
+	serviceEntry := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      model.ServiceEntry.Type,
+			Group:     model.ServiceEntry.Group + model.IstioAPIGroupDomain,
+			Version:   model.ServiceEntry.Version,
+			Name:      fmt.Sprintf("service-entry-%s", remoteService.host),
+			Namespace: namespace,
+			// Annotations: annotations(config), //  TODO (MB)
+		},
+		Spec: &v1alpha3.ServiceEntry{
+			Hosts: []string{remoteService.host},
```

Comment:
This needs to be of the form `service.namespace.global`. Maybe the user should pass in an optionally qualified name (namespace is `default` if not specified). So users can pass `reviews.ns1:9080` or `reviews:9080` (which is the same as `reviews.default:9080`). Then here you would append `.global` before setting it.

Reply:
Done